### PR TITLE
chore: add cdn certificate to cloudfront and enable feature

### PR DESCRIPTION
**Pull request by CaptainCarpensir**
This adds `cdn.certificate` which allows the user to use certificates imported in `us-east-1` with CloudFront.
There's a few more edge cases needed for the feature to be release ready namely to work with or invalidate `hosted_zone` field in svc manifest with CloudFront.

@iamhopaul123 will be adding the capability for CloudFront to read aliases from the stack parameters to allow a CloudFront distribution creation with imported certs or `--domain` to be able to occur entirely within Copilot

Nearly completes milestone 1 of #3701 and requires #3842 to be merged before merging this

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/cli/deploy/svc.go [hunk 1]
```diff
@@ -1390,12 +1396,20 @@ func (d *lbWebSvcDeployer) validateALBRuntime() error {
 	}
 	if hasImportedCerts {
 		aliases, err := d.lbMft.RoutingRule.Alias.ToStringSlice()
+		cdnCert := d.environmentConfig.CDNConfig.CDNConfig.Certificate
+		defaultValidator := d.newAliasCertValidator(nil)
+		cfValidator := d.newAliasCertValidator(aws.String("us-east-1")) // make us-east-1 a constant somewhere
```
Comment (dannyrandall) [inline internal/pkg/cli/deploy/svc.go]: let's make it a constant somewhere :)

File: internal/pkg/cli/deploy/svc.go [hunk 2]
```diff
@@ -1390,12 +1396,20 @@ func (d *lbWebSvcDeployer) validateALBRuntime() error {
...
 		if err != nil {
 			return fmt.Errorf("convert aliases to string slice: %w", err)
 		}
```
Comment (dannyrandall) [inline internal/pkg/cli/deploy/svc.go]:
```suggestion
		aliases, err := d.lbMft.RoutingRule.Alias.ToStringSlice()
		if err != nil {
			return fmt.Errorf("convert aliases to string slice: %w", err)
		}
		
		cdnCert := d.environmentConfig.CDNConfig.CDNConfig.Certificate
		defaultValidator := d.newAliasCertValidator(nil)
		cfValidator := d.newAliasCertValidator(aws.String("us-east-1")) // make us-east-1 a constant somewhere
```

File: internal/pkg/template/templates/environment/cf.yml [hunk 3]
```diff
@@ -127,7 +127,7 @@ Resources:
 {{- end}}
       Tags:
         - Key: Name
-          Value: !Sub 'copilot-${AppName}-${EnvironmentName}-lb'
+          Value: !Sub 'copilot-${AppName}-${EnvironmentName}-lb-http'
```
Comment (dannyrandall) [inline internal/pkg/template/templates/environment/cf.yml:130]: ❤️

File: internal/pkg/template/templates/environment/partials/environment-manager-role.yml [hunk 4]
```diff
@@ -32,6 +32,11 @@ EnvironmentManagerRole:
 {{- range $arn := .PrivateHTTPConfig.ImportedCertARNs}}
           - "{{$arn}}"
 {{- end}}
+{{- if .CDNConfig}}
+{{- if .CDNConfig.ViewerCertificate}}
```
Comment (paragbhingre) [inline internal/pkg/template/templates/environment/partials/environment-manager-role.yml]:
```suggestion
{{- if and .CDNConfig .CDNConfig.ViewerCertificate}}
```

You can do this instead of two if statements

File: internal/pkg/cli/deploy/svc.go (hunk 1 above)
Comment (iamhopaul123) [inline internal/pkg/cli/deploy/svc.go]: we can make it a constant here for now.

File: internal/pkg/cli/deploy/svc.go [hunk 5]
```diff
@@ -327,10 +327,10 @@ type customResourcesFunc func(fs template.Reader) ([]*customresource.CustomResou
 type lbWebSvcDeployer struct {
 	*svcDeployer
 	appVersionGetter       versionGetter
-	aliasCertValidator     aliasCertValidator
 	publicCIDRBlocksGetter publicCIDRBlocksGetter
 	lbMft                  *manifest.LoadBalancedWebService
 	customResources        customResourcesFunc
+	newAliasCertValidator  func(region *string) aliasCertValidator
```
Comment (iamhopaul123) [inline internal/pkg/cli/deploy/svc.go]:
```suggestion
	newAliasCertValidator  func(optionalRegion *string) aliasCertValidator
```

File: internal/pkg/cli/deploy/svc.go [hunk 6]
```diff
@@ -1390,12 +1396,20 @@ func (d *lbWebSvcDeployer) validateALBRuntime() error {
 	}
 	if hasImportedCerts {
 		aliases, err := d.lbMft.RoutingRule.Alias.ToStringSlice()
+		cdnCert := d.environmentConfig.CDNConfig.CDNConfig.Certificate
+		defaultValidator := d.newAliasCertValidator(nil)
```
Comment (iamhopaul123) [inline internal/pkg/cli/deploy/svc.go]: How about `albCertValidator` and `cfCertValidator`

File: internal/pkg/cli/deploy/svc.go [hunk 7]
```diff
@@ -1390,12 +1396,20 @@ func (d *lbWebSvcDeployer) validateALBRuntime() error {
...
+		cfValidator := d.newAliasCertValidator(aws.String("us-east-1")) // make us-east-1 a constant somewhere
 		if err != nil {
 			return fmt.Errorf("convert aliases to string slice: %w", err)
 		}
-		if err := d.aliasCertValidator.ValidateCertAliases(aliases, d.environmentConfig.HTTPConfig.Public.Certificates); err != nil {
+		if err := defaultValidator.ValidateCertAliases(aliases, d.environmentConfig.HTTPConfig.Public.Certificates); err != nil {
 			return fmt.Errorf("validate aliases against the imported certificate for env %s: %w", d.env.Name, err)
```
Comment (iamhopaul123) [inline internal/pkg/cli/deploy/svc.go]:
```suggestion
			return fmt.Errorf("validate aliases against the imported public ALB certificate for env %s: %w", d.env.Name, err)
```

File: internal/pkg/cli/deploy/svc.go [hunk 8]
```diff
@@ -1390,12 +1396,20 @@ func (d *lbWebSvcDeployer) validateALBRuntime() error {
...
 		}
+		if cdnCert != nil {
+			if err := cfValidator.ValidateCertAliases(aliases, []string{*cdnCert}); err != nil {
+				return fmt.Errorf("validate aliases against the cdn imported certificate for env %s: %w", d.env.Name, err)
```
Comment (iamhopaul123) [inline internal/pkg/cli/deploy/svc.go]:
```suggestion
				return fmt.Errorf("validate aliases against the imported CDN certificate for env %s: %w", d.env.Name, err)
```

File: internal/pkg/manifest/validate_env.go [hunk 9]
```diff
@@ -326,6 +338,11 @@ func (i RestrictiveIngress) validate() error {
 
 // validate is a no-op for AdvancedCDNConfig.
```
Comment (paragbhingre) [inline internal/pkg/manifest/validate_env.go]: You might want to change this description of the validate method.

File: internal/pkg/deploy/cloudformation/stack/env.go [hunk 10]
```diff
@@ -353,7 +354,15 @@ func (e *BootstrapEnvStackConfig) ToEnv(stack *cloudformation.Stack) (*config.En
...
 
 func (e *EnvStackConfig) cdnConfig() *template.CDNConfig {
-	return nil // no-op - return &template.CDNConfig{} when feature is ready
+	if e.in.Mft == nil {
+		return nil
+	}
+	if !e.in.Mft.CDNConfig.CDNEnabled() {
+		return nil
+	}
+	return &template.CDNConfig{
```
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/stack/env.go:363]: woohoo it is happening

File: internal/pkg/manifest/validate_env.go [hunk 11]
```diff
@@ -19,7 +20,7 @@ var (
 // Validate returns nil if Environment is configured correctly.
 func (e Environment) Validate() error {
 	if err := e.EnvironmentConfig.validate(); err != nil {
-		return fmt.Errorf(`validate "network": %w`, err)
+		return fmt.Errorf(`validate environment: %w`, err)
```
Comment (iamhopaul123) [inline internal/pkg/manifest/validate_env.go]: We don't even need to wrap this error i think. because otherwise the error msg will be

```
validate environment manifest for "test": validate environment: some error"
```

File: internal/pkg/manifest/validate_env.go [hunk 12]
```diff
@@ -41,6 +42,17 @@ func (e EnvironmentConfig) validate() error {
 	if e.IsIngressRestrictedToCDN() && !e.CDNConfig.CDNEnabled() {
 		return errors.New("CDN must be enabled to limit security group ingress to CloudFront")
 	}
+	if e.CDNConfig.CDNEnabled() {
+		if e.HTTPConfig.Public.Certificates == nil {
+			if e.CDNConfig.CDNConfig.Certificate != nil {
```
Comment (iamhopaul123) [inline internal/pkg/manifest/validate_env.go]: We should probably rename the second CDNConfig in `environmentCDNConfig` to just `Config`, since the struct name should have the namespace info.

File: internal/pkg/manifest/validate_env.go [hunk 13]
```diff
@@ -41,6 +42,17 @@ func (e EnvironmentConfig) validate() error {
...
+				return errors.New("must import ALB certificates to set \"cdn.certificate\"")
```
Comment (iamhopaul123) [inline internal/pkg/manifest/validate_env.go]: We might want to provide more context to users saying they have to specify together. Can we make use of `errFieldMustBeSpecified`?

File: internal/pkg/manifest/validate_env.go [hunk 14]
```diff
@@ -41,6 +42,17 @@ func (e EnvironmentConfig) validate() error {
...
+			}
+		} else {
+			if e.CDNConfig.CDNConfig.Certificate == nil {
+				return errors.New("must set \"cdn.certificate\" to import ALB certificates with CDN enabled")
```
Comment (iamhopaul123) [inline internal/pkg/manifest/validate_env.go]: same for this one

**Comment by iamhopaul123** [comment]
Also we need to write to the alternate name for CF distribution right? Will that be in the following PR?

**Comment by CaptainCarpensir** [comment]
> Also we need to write to the alternate name for CF distribution right? Will that be in the following PR?

Ah I completely forgot about this! At this point yeah it should be, I don't think it will take much time, I'm just worried that we're enabling the feature when it's still a little too incomplete if that's the case. After this there are two loose ends to wrap up:
- alternate domain name writing for cloudfront
- validation of some kind for services using the `hosted_zone` feature

**Comment by iamhopaul123** [comment]
I don't mind validation to go after this PR but yeah I agree the alternate name setting we might merge that one prior to this feat enabling PR.

**Comment by CaptainCarpensir** [comment]
Alright, this PR isn't very big on its own so I'll add that in here then re-enable for reviews 👍

File: internal/pkg/manifest/validate_env.go [hunk 15]
```diff
@@ -41,6 +42,17 @@ func (e EnvironmentConfig) validate() error {
...
+			}
+		}
+	}
```
Comment (paragbhingre) [inline internal/pkg/manifest/validate_env.go]: This condition has multiple nested ifs, I was thinking what we can do with it. I tried doing few permutations with the if logic and it passes your tests in validate_env but fails few other tests for some weird reasons. 
We can at least take this `e.CDNConfig.CDNConfig.Certificate` out in a variable and make it more readable.

**Review by paragbhingre — COMMENTED** [review]
Looks good🤩 , just few nits.

File: internal/pkg/manifest/validate_env.go (hunk 12 above)
Comment (paragbhingre) [reply internal/pkg/manifest/validate_env.go]: +100

File: internal/pkg/manifest/errors.go [hunk 16]
```diff
@@ -56,16 +56,21 @@ func (e *ErrUnknownProvider) Is(target error) bool {
...
 func (e *errFieldMustBeSpecified) Error() string {
 	errMsg := fmt.Sprintf(`%q must be specified`, e.missingField)
 	if len(e.conditionalFields) == 0 {
 		return errMsg
 	}
-	return fmt.Sprintf(`%s if %s %s specified`, errMsg, english.WordSeries(quoteStringSlice(e.conditionalFields), "or"),
+	conjunction := "or"
+	if e.allMustBeSpecified {
+		conjunction = "and"
+	}
```
Comment (dannyrandall) [inline internal/pkg/manifest/errors.go:72]: idk why but this makes me laugh

File: internal/pkg/manifest/validate_env_test.go [hunk 17]
```diff
@@ -695,14 +719,27 @@ func TestCDNConfiguration_validate(t *testing.T) {
...
 			},
+			wantedErrorMsgPrefix: "parse cdn certificate:",
 		},
 	}
 	for name, tc := range testCases {
 		t.Run(name, func(t *testing.T) {
 			gotErr := tc.in.validate()
-			if tc.wantedError != nil {
+			if tc.wantedErrorMsgPrefix != "" {
+				require.Error(t, gotErr)
```
Comment (dannyrandall) [inline internal/pkg/manifest/validate_env_test.go:739]: this is covered by the check below
```suggestion
```

**Review by dannyrandall — APPROVED** [review]

File: internal/pkg/deploy/cloudformation/stack/env.go (hunk 10 above)
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/stack/env.go:363]: We should probably revert this change.

**Review by iamhopaul123 — APPROVED** [review]
